feat(experimental): Initialize Android SDK from json configuration

antonis opened the pull request:
## :loudspeaker: Type of change
<!--- Put an `x` in the boxes that apply -->
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring

**Depends on https://github.com/getsentry/sentry-react-native/pull/4445 and https://github.com/getsentry/sentry-react-native/pull/4479**

## :scroll: Description
<!--- Describe your changes in detail -->
Adds the following methods that allow initialisation of Sentry Android in RN applications:
- `RNSentrySDK.init(Context context)`: Initialises using the `sentry.options.json` file located in the root of the RN project.
- `RNSentrySDK.init(Context context, Sentry.OptionsConfiguration<SentryAndroidOptions> configuration)`: Initialises with the passed `configuration` and the `sentry.options.json` if it exists.

## :bulb: Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Part of https://github.com/getsentry/sentry-react-native/issues/3608

## :green_heart: How did you test it?
CI, Manual testing

## :pencil: Checklist
<!--- Put an `x` in the boxes that apply -->
- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled
- [ ] I updated the docs if needed.
- [ ] I updated the wizard if needed.
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps

krystofwoldrich commented on packages/core/android/src/main/java/io/sentry/react/RNSentrySDK.java:
```diff
@@ -0,0 +1,82 @@
+   * @param context Android Context
+   * @param options Map with options
+   */
+  public static void startWithOptions(
+      @NotNull final Context context, @NotNull final Map<String, Object> options) {
```
Let's match the `SentryAndroid` init signature, so the RN manual init can be a drop in replacement.

https://github.com/getsentry/sentry-java/blob/160762108453416271f7135b4e1d4dfc329fe871/sentry-android-core/src/main/java/io/sentry/android/core/SentryAndroid.java#L67-L77

antonis replied:
> This public interface should allow user to overwrite/extend the options specified in the sentry.options.json file using the configuration handler.

I think this is a great idea and makes the exposed api really powerful allowing for dynamic configurations that a json file cannot provide. Updated with 25d142a3ca9653c5da2cb4a8f98c7a29bab48309

krystofwoldrich commented on packages/core/android/src/main/java/io/sentry/react/RNSentrySDK.java:
```diff
@@ -0,0 +1,82 @@
+   *
+   * @param context Android Context
+   */
+  public static void start(@NotNull final Context context) {
```
I'm thinking about using init here to match the `SentryAndroid`. Cocoa uses `start` because `init` is reserved for constructors in Obj-C.

antonis replied:
Updated with 7788b81730b579fe8ebe2013abec4c01f31ee347

krystofwoldrich commented on packages/core/android/src/main/java/io/sentry/react/RNSentryMapConverter.java:
```diff
@@ -131,4 +136,37 @@ private static void addValueToWritableMap(WritableMap writableMap, String key, O
+    return JavaOnlyMap.of(keysAndValues);
+  }
+
+  private static Map<String, Object> jsonObjectToMap(JSONObject jsonObject) {
+    Map<String, Object> map = new HashMap<>();
+    Iterator<String> keys = jsonObject.keys();
+    while (keys.hasNext()) {
+      String key = keys.next();
+      Object value = null;
+      try {
+        value = jsonObject.get(key);
+      } catch (JSONException e) {
+        throw new RuntimeException(e);
+      }
+      map.put(key, value);
+    }
+    return map;
+  }
```
We should check with the Android team, this might be already implemented in SentryAndroid.

krystofwoldrich commented on packages/core/android/src/main/java/io/sentry/react/RNSentryMapConverter.java:
```diff
@@ -131,4 +136,37 @@ private static void addValueToWritableMap(WritableMap writableMap, String key, O
+  }
+
+  public static ReadableMap mapToReadableMap(Map<String, Object> map) {
+    // We are not directly using `convertToWritable` since `Arguments.createArray()`
+    // fails before bridge initialisation
```
Can you share more details? I didn't know that, is there an issue that we could link? Or is that expected behaviour?

antonis replied:
The implementation would have been simpler if we could replace the `mapToReadableMap(map)` above with `(WritableMap) convertToWritable(map)` but this results in an `ExceptionInInitializerError` when called from the Android Application `onCreate` method before RN initialises. I think the root of the issue is similar with [this old RN issue](https://github.com/facebook/react-native/issues/10352) or [this one](https://github.com/facebook/react-native/issues/10379). 

<details><summary>Full stack trace</summary>
<p>

```
java.lang.ExceptionInInitializerError
	at com.facebook.react.bridge.Arguments.createMap(Arguments.java:163)
	at io.sentry.react.RNSentryMapConverter.convertToWritable(RNSentryMapConverter.java:39)
	at io.sentry.react.RNSentryMapConverter.jsonObjectToReadableMap(RNSentryMapConverter.java:142)
	at io.sentry.react.RNSentrySDK.start(RNSentrySDK.java:50)
	at io.sentry.reactnative.sample.MainApplication.initializeSentry(MainApplication.kt:59)
	at io.sentry.reactnative.sample.MainApplication.onCreate(MainApplication.kt:42)
	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1388)
	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:7586)
	at android.app.ActivityThread.-$$Nest$mhandleBindApplication(Unknown Source:0)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2449)
	at android.os.Handler.dispatchMessage(Handler.java:109)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at android.app.ActivityThread.main(ActivityThread.java:8787)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:591)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:871)
Caused by: java.lang.IllegalStateException: SoLoader.init() not yet called
	at com.facebook.soloader.SoLoader.assertInitialized(SoLoader.java:1273)
	at com.facebook.soloader.SoLoader.loadLibraryOnNonAndroid(SoLoader.java:880)
	at com.facebook.soloader.SoLoader.loadLibrary(SoLoader.java:825)
	at com.facebook.soloader.SoLoader.loadLibrary(SoLoader.java:812)
	at com.facebook.react.bridge.ReactBridge.staticInit(ReactBridge.java:34)
	at com.facebook.react.bridge.NativeMap.<clinit>(NativeMap.kt:20)
	at com.facebook.react.bridge.Arguments.createMap(Arguments.java:163) 
	at io.sentry.react.RNSentryMapConverter.convertToWritable(RNSentryMapConverter.java:39) 
	at io.sentry.react.RNSentryMapConverter.jsonObjectToReadableMap(RNSentryMapConverter.java:142) 
	at io.sentry.react.RNSentrySDK.start(RNSentrySDK.java:50) 
	at io.sentry.reactnative.sample.MainApplication.initializeSentry(MainApplication.kt:59) 
	at io.sentry.reactnative.sample.MainApplication.onCreate(MainApplication.kt:42) 
	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1388) 
	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:7586) 
	at android.app.ActivityThread.-$$Nest$mhandleBindApplication(Unknown Source:0) 
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2449) 
	at android.os.Handler.dispatchMessage(Handler.java:109) 
	at android.os.Looper.loopOnce(Looper.java:232) 
	at android.os.Looper.loop(Looper.java:317) 
	at android.app.ActivityThread.main(ActivityThread.java:8787) 
	at java.lang.reflect.Method.invoke(Native Method) 
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:591) 
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:871) 
```
</p>
</details>

antonis replied:
Iterating on this calling `initializeSentry()` after `SoLoader.init` fixes the above issue. 

https://github.com/getsentry/sentry-react-native/blob/2dc8f2246d81880f9b42e6c19942cc94a46bffed/samples/react-native/android/app/src/main/java/io/sentry/reactnative/sample/MainApplication.kt#L41-L50

We can probably drop the custom implementation in favour of reusing our existing `convertToWritable` documenting the requirement. Wdyt @krystofwoldrich ?

krystofwoldrich replied:
We can't do that, because we need to be able to initialize before any RN code. To capture any potential errors/crashes.

antonis replied:
> because we need to be able to initialize before any RN code

Makes sense. I'll stick with the new `mapToReadableMap` implementation that does not depend on RN.

krystofwoldrich replied:
But I think we can refactor `convertToWritable` to be more reusable.

We could create `convertToNativeWritable` and `convertToJavaWritable`.

Which internally use adjusted `convertToWritable` which depending on the need uses `Arguments.createArray()` or `new JavaOnlyArray`.

---

`Arguments.createArray()` and `new JavaOnlyArray` implement the same interface, but the `createArray` instantiates the cpp implemented array which can be passed over the bridge to JS. The `JavaOnly*` classes can't be passed to JS.

antonis replied:
Sounds like a good idea 🙇  I'll pursue this direction and iterate back.

antonis replied:
Since this change involved many modifications, I spun it off into a separate PR https://github.com/getsentry/sentry-react-native/pull/4469

krystofwoldrich commented on packages/core/RNSentryAndroidTester/app/src/androidTest/java/io/sentry/rnsentryandroidtester/RNSentryMapConverterTest.kt:
```diff
@@ -359,4 +363,40 @@ class MapConverterTest {
         assertEquals(actual, expectedMap1)
     }
+
+    @Test
+    fun testJsonObjectToReadableMap() {
+        val json =
+            JSONObject().apply {
+                put("stringKey", "stringValue")
+                put("booleanKey", true)
+                put("intKey", 123)
+            }
+
+        val result = RNSentryMapConverter.jsonObjectToReadableMap(json)
+
+        assertNotNull(result)
+        assertTrue(result is JavaOnlyMap)
+        assertEquals("stringValue", result.getString("stringKey"))
+        assertEquals(true, result.getBoolean("booleanKey"))
+        assertEquals(123, result.getInt("intKey"))
+    }
+
+    @Test
+    fun testMapToReadableMap() {
+        val map =
+            mapOf(
+                "stringKey" to "stringValue",
+                "booleanKey" to true,
+                "intKey" to 123,
+            )
+
+        val result = RNSentryMapConverter.mapToReadableMap(map)
+
+        assertNotNull(result)
+        assertTrue(result is JavaOnlyMap)
+        assertEquals("stringValue", result.getString("stringKey"))
+        assertEquals(true, result.getBoolean("booleanKey"))
+        assertEquals(123, result.getInt("intKey"))
+    }
```
We should also ensure nested maps and arrays are handled because some Sentry init options require those types.

antonis replied:
Makes sense 👍 

The nested maps/arrays should be handled with the https://github.com/getsentry/sentry-react-native/pull/4469 enhancement/refactoring discussed above https://github.com/getsentry/sentry-react-native/pull/4451#discussion_r1923360541

I'll also extend the json handling to support nested structures with https://github.com/getsentry/sentry-react-native/pull/4470

krystofwoldrich commented on packages/core/android/src/main/java/io/sentry/react/RNSentrySDK.java:
```diff
@@ -0,0 +1,82 @@
+   */
+  public static void start(@NotNull final Context context) {
+    try {
+      JSONObject jsonObject = getOptionsFromConfigurationFile(context);
+      ReadableMap rnOptions = RNSentryMapConverter.jsonObjectToReadableMap(jsonObject);
```
Our current lint doesn't enforce it, but let's use the `@NotNull/@Nullable` annotations here.

Let's add tests to ensure we gracefully continue if the file is not present or contains broken JSON and the init with configure callback is used.

If the `start()` is used I agree it make sense to fail as there is no other way to add the options.

antonis replied:
Good point 👍 
I've handled this in https://github.com/getsentry/sentry-react-native/pull/4474

antonis commented:
Taking a note to recheck the initialisation defaults separately. Those should be [aligned with the iOS implementation](https://github.com/getsentry/sentry-react-native/pull/4444/files#diff-90d9676a679f04d29edee62e5422c73f60a04ad890ab1d13191826912466e9d4R103).

krystofwoldrich commented on samples/react-native/android/app/src/main/java/io/sentry/reactnative/sample/MainApplication.kt:
```diff
@@ -51,9 +51,8 @@ class MainApplication :
+            // Options set here will apply to the Android SDK overriding the ones from `sentry.options.json`
             options.dsn = "https://1df17bd4e543fdb31351dee1768bb679@o447951.ingest.sentry.io/5428561"
             options.isDebug = true
 
```
The `beforeSend` callback should not be needed anymore, the `RNSentrySDK.init` should handle the RN defaults internally.

antonis replied:
Removed with 7522fcebe114a06f11a9421a96539904e4edda2b

antonis commented on packages/core/sentry.gradle:
```diff
@@ -17,7 +17,52 @@ project.ext.shouldSentryAutoUpload = { ->
+                androidAssetsDir.mkdirs()
+            }
+
+            copy {
```
Yep, it always replaces any existing file with that name in the assets. Note that after the build the file is deleted from assets.

krystofwoldrich commented on packages/core/RNSentryAndroidTester/app/src/test/java/io/sentry/rnsentryandroidtester/RNSentryCompositeOptionsConfigurationTest.kt:
```diff
@@ -0,0 +1,51 @@
+            OptionsConfiguration<SentryAndroidOptions> { options ->
+                options.dsn = "https://base-dsn@sentry.io"
+                options.isDebug = false
+                options.release = "some-release"
```
I think `options.environment = "base"` is missing here.

krystofwoldrich replied:
I took a look at it again and I now understand the different scenarios, it's good as is.

krystofwoldrich commented on packages/core/android/src/main/java/io/sentry/react/RNSentryStart.java:
```diff
@@ -188,6 +215,42 @@ static void getSentryAndroidOptions(
+    BeforeSendCallback userBeforeSend = options.getBeforeSend();
+    options.setBeforeSend(
+        (event, hint) -> {
+          // Unhandled JS Exception are processed by the SDK on JS layer
+          // To avoid duplicates we drop them in the native SDKs
+          if (event.getExceptions() != null && !event.getExceptions().isEmpty()) {
+            String exType = event.getExceptions().get(0).getType();
+            if (exType != null && exType.contains("Unhandled JS Exception")) {
+              return null; // Skip sending this event
+            }
+          }
```
We can omit this JS Exception check because it's set in 

https://github.com/getsentry/sentry-react-native/pull/4451/files#diff-7af7f42daf65903fec96e896fc2518a5422f8545b9bc70c9e512e4da060a589dR199

which should be moved to the RN defaults.

antonis replied:
Good catch @krystofwoldrich 👍 
Updated with 821fef94291860ba463269804a57016e3a3dec3f

krystofwoldrich commented:
I think [this](https://github.com/getsentry/sentry-react-native/pull/4451/files#diff-7af7f42daf65903fec96e896fc2518a5422f8545b9bc70c9e512e4da060a589dR211) should be moved to ReactDefaults.

krystofwoldrich commented:
[The SDK name and version](https://github.com/getsentry/sentry-react-native/pull/4451/files#diff-7af7f42daf65903fec96e896fc2518a5422f8545b9bc70c9e512e4da060a589dR85-R97) should be moved to ReactDefaults.

krystofwoldrich commented on packages/core/RNSentryAndroidTester/app/src/androidTest/java/io/sentry/react/RNSentrySDKTest.kt:
```diff
@@ -0,0 +1,127 @@
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class RNSentrySDKTest {
```
These integration tests look great. I like that they are close to what the user will execute.

Can you please add similar tests for the RN defaults and finals? Ideally, we check that the values are set/overwritten as expected both when reading options from the file and when the options are set directly in the callback.

antonis replied:
Sounds like a good idea. I'll iterate on this 👍

antonis replied:
I've added some basic tests for the defaults and finals with 2b7a319efc391f4c403e7482ccaedb728306861f and added `RNSentrySDK` tests with ced915d5593eae951f38779c9dfde22abcb58305 verifying the default/final configuration and the overriding behaviour. Note that a minor refactoring was needed d236c82480960c22e96deac2da4bfa8ac6c49588 since it was not possible to mock the Sentry SDK for Android testing.

krystofwoldrich commented:
I think the PR is almost done, I've just noticed two items that should also be moved to the react defaults.

antonis commented:
Thank you for the feedback Krystof 🙇 

> [The SDK name and version](https://github.com/getsentry/sentry-react-native/pull/4451/files#diff-7af7f42daf65903fec96e896fc2518a5422f8545b9bc70c9e512e4da060a589dR85-R97) should be moved to ReactDefaults. So the correct name is set in both init with file and without file.

Updated with 43b26db87198ec2e7dbd6c60046b6783d4230eea

> I think [this](https://github.com/getsentry/sentry-react-native/pull/4451/files#diff-7af7f42daf65903fec96e896fc2518a5422f8545b9bc70c9e512e4da060a589dR211) should be moved to ReactDefaults. So the current activity is set regardless of the init with file or init without the file.

Updated with d2fd58fe76b3143d723dd4867969fea22532e1a8

antonis commented on packages/core/android/src/main/java/io/sentry/react/RNSentryStart.java:
```diff
@@ -27,19 +29,52 @@
+        options -> getSentryAndroidOptions(options, rnOptions, null, logger);
+    RNSentryCompositeOptionsConfiguration compositeConfiguration =
+        new RNSentryCompositeOptionsConfiguration(
+            RNSentryStart::updateWithReactDefaults,
```
Updated with 5b9665aefe861f4b4ac8782b4a8e80b210c6c6c3

krystofwoldrich commented:
I've noticed the E2E tests started failing on this branch, is it possibly caused by some of the init changes?

antonis commented:
> I've noticed the E2E tests started failing on this branch, is it possibly caused by some of the init changes?

I think it is CI related since it started with a [small internal change](https://github.com/getsentry/sentry-react-native/pull/4451/commits/821fef94291860ba463269804a57016e3a3dec3f) and the errors seems to be [related with Maestro](https://github.com/getsentry/sentry-react-native/actions/runs/13118516701/job/36614276429?pr=4451). I also noticed [another deps PR](https://github.com/getsentry/sentry-react-native/actions/runs/13113646994/job/36585403940?pr=4503) with similar errors.
I'll check again and iterate back.

krystofwoldrich reviewed (approved):
Thank you! 🚀
